### PR TITLE
(proof-new) Add datatypes proof checker

### PR DESCRIPTION
**Pull request by ajreynol**
This adds the proof rules and proof checker for datatypes.

### CODE REVIEW

File: src/expr/proof_rule.h [hunk 1]
```diff
@@ -674,6 +674,49 @@ enum class PfRule : uint32_t
   // ---------------------
   // Conclusion: F
   ARRAYS_TRUST,
+  
+  //================================================= Datatype rules
+  // ======== Unification
+  // Children: (P:(= (C t1 ... tn) (C s1 ... sn)))
+  // Arguments: (i)
+  // ----------------------------------------
+  // Conclusion: (= ti si)
```
Comment (nafur) [inline src/expr/proof_rule.h:683]: Could you add a bit of general information here? I guess it mostly boils done to a very basic explanation of what DT is all about. I guess `C` stands for "Constructor"?

File: src/theory/datatypes/proof_checker.cpp [hunk 2]
```diff
@@ -0,0 +1,135 @@
...
+        || children[0][1].getKind() != kind::APPLY_CONSTRUCTOR
+        || children[0][0].getOperator() != children[0][1].getOperator()
+        || !getUInt32(args[0], i))
+    {
+      return Node::null();
+    }
+    if (i >= children[0][0].getNumChildren())
+    {
+      return Node::null();
+    }
```
Comment (nafur) [inline src/theory/datatypes/proof_checker.cpp:56]: Why did you put the last condition in a separate if?

**Review by nafur — APPROVED** [review]
LGTM, but I'd really like a bit more documentation on the proof rules.

File: src/theory/datatypes/proof_checker.cpp (hunk 2 above)
Comment (ajreynol) [reply src/theory/datatypes/proof_checker.cpp:56]: To me it gets a little weird following the code if one condition above (`getUInt32(args[0], i)`) is responsible for setting `i` and then another condition uses it.  Should work but relies on order of operations within an if-condition, anyways, for clarity I've broken at that point.

Comment (nafur) [reply src/theory/datatypes/proof_checker.cpp:56]: I see. So first of: this is safe as the evaluation order for `||` and `&&` is left-to-right (to allow for short-circuiting, in contrast to most other operator / function evaluations). If we find this confusing, we should find the first condition to be confusing as well, though, as `children[0].getKind() != kind::EQUAL` ensures that the access to `children[0][0]` is safe.
No strong opinion on my side, though...

**Pull request merged**